src/tests/testMath.cc

```diff
@@ -266,3 +279,151 @@ TestMath::testNaturalSum()
+template <typename T, typename U, typename P = T>
+static P
+GoodProduct(const T t, const U u)
+{
+    return SuccessProductTester<P>::Test(t, u);
+}
+
+void
+TestMath::testIntegralProduct()
+{
+    // negative parameters are banned in any position
+    TestOverflowForEitherMultiplicationType(min64s, zero8s);
+    TestOverflowForEitherMultiplicationType(min64s, zero8u);
+    TestOverflowForEitherMultiplicationType(min64s, max64s);
+    TestOverflowForEitherMultiplicationType(min64s, max64u);
+    TestOverflowForEitherMultiplicationType(min8s, zero8s);
+    TestOverflowForEitherMultiplicationType(min8s, zero8s);
+    TestOverflowForEitherMultiplicationType(min8s, zero8u);
+    TestOverflowForEitherMultiplicationType(min8s, max64s);
+    TestOverflowForEitherMultiplicationType(min8s, max64u);
+    TestOverflowForEitherMultiplicationType(-1, -1);
+    TestOverflowForEitherMultiplicationType(-1, zero8s);
+    TestOverflowForEitherMultiplicationType(-1, zero8u);
+    TestOverflowForEitherMultiplicationType(-1, max64s);
+    TestOverflowForEitherMultiplicationType(-1, max64u);
+    TestOverflowForEitherMultiplicationType(-1, one8s);
+    TestOverflowForEitherMultiplicationType(-1, one8u);
+    TestOverflowForEitherMultiplicationType(-1, one64s);
+    TestOverflowForEitherMultiplicationType(-1, one64u);
+
+    // these overflow regardless of which parameter determines the product type
+    TestOverflowForEitherMultiplicationType(max8u, two8u);
+    TestOverflowForEitherMultiplicationType(max8u, two8s);
+    TestOverflowForEitherMultiplicationType(max8u, max8s);
+    TestOverflowForEitherMultiplicationType(max8s, two8s);
+    TestOverflowForEitherMultiplicationType(max64u, two8u);
+    TestOverflowForEitherMultiplicationType(max64u, two8s);
+    TestOverflowForEitherMultiplicationType(max64u, two64u);
+    TestOverflowForEitherMultiplicationType(max64u, two64s);
+    TestOverflowForEitherMultiplicationType(max64u, max64s);
+    TestOverflowForEitherMultiplicationType(max64s, two8u);
+    TestOverflowForEitherMultiplicationType(max64s, two8s);
+    TestOverflowForEitherMultiplicationType(max64s, two64s);
+
+    // these overflow only if the second parameter determines the product type
+    TestSuccessForFirstMultiplicationType(max8u, one8s);
+    TestSuccessForFirstMultiplicationType(max64u, one8s);
+    TestSuccessForFirstMultiplicationType(max64u, one8u);
+    TestSuccessForFirstMultiplicationType(max64u, one64s);
+    TestSuccessForFirstMultiplicationType(max64s, one8s);
+    TestSuccessForFirstMultiplicationType(max64s, one8u);
+    TestSuccessForFirstMultiplicationType(max64u, zero8u);
+    TestSuccessForFirstMultiplicationType(max64u, zero8s);
+    TestSuccessForFirstMultiplicationType(max64u, zero64s);
+
+    // a few products with known values
+    CPPUNIT_ASSERT_EQUAL(zero8s, GoodProduct(zero8s, zero8u));
+    CPPUNIT_ASSERT_EQUAL(zero64s, GoodProduct(zero64s, zero64u));
+    CPPUNIT_ASSERT_EQUAL(2, GoodProduct(2, 1));
```

Comment:
Let's add one more unity test:
```suggestion
    CPPUNIT_ASSERT_EQUAL(one8u, GoodProduct(one8u, one64s));
    CPPUNIT_ASSERT_EQUAL(2, GoodProduct(2, 1));
```